aten/src/ATen/native/FractionalMaxPool3d.cpp

```diff
@@ -0,0 +1,420 @@
+
+    for (h = 0; h < outputH; ++h) {
+      int inputHStart = sequenceH[h];
+
+      for (w = 0; w < outputW; ++w) {
+        int inputWStart = sequenceW[w];
+
+        for (t = 0; t < outputT; ++t) {
+          int inputTStart = sequenceT[t];
+
+          scalar_t maxVal = -std::numeric_limits<scalar_t>::infinity();
+          int maxIndex = -1;
+
+          int h2, w2, t2;
+          for (h2 = inputHStart; h2 < inputHStart + poolSizeH; ++h2) {
+            for (w2 = inputWStart; w2 < inputWStart + poolSizeW; ++w2) {
+              for (t2 = inputTStart; t2 < inputTStart + poolSizeT; ++t2) {
+                AT_ASSERT(h2 >= 0 && h2 < inputH);
+                AT_ASSERT(w2 >= 0 && w2 < inputW);
+                AT_ASSERT(t2 >= 0 && t2 < inputT);
+
+                int planeIndex = h2 * inputW * inputT + w2 * inputT + t2;
+                scalar_t val = inputForPlane[planeIndex];
+                if (val > maxVal) {
+                  maxVal = val;
+                  maxIndex = planeIndex;
+                }
+              }
+            }
+          }
+
+          AT_ASSERT(maxVal != -std::numeric_limits<scalar_t>::infinity());
+          AT_ASSERT(maxIndex != -1);
+
+          outputForPlane[h * outputW * outputT + w * outputT + t] = maxVal;
+          indicesForPlane[h * outputW * outputT + w * outputT + t] = maxIndex;
+        }
+      }
+    }
+
+  }
+}
+
+template<typename scalar_t>
+static void fractional_max_pool3d_out_frame(
+  scalar_t* input,
+  scalar_t* output,
+  int64_t* indices,
+  scalar_t* randomSamples,
+  int numBatch, int numPlanes,
+  int inputT, int inputW, int inputH,
+  int outputT, int outputW, int outputH,
+  int poolSizeT, int poolSizeW, int poolSizeH) {
+    if(numBatch == 1) {
+      fractional_max_pool3d_out_single_batch_frame<scalar_t>(
+        input, output, indices, randomSamples,
+        numPlanes,
+        inputT, inputW, inputH,
+        outputT, outputW, outputH,
+        poolSizeT, poolSizeW, poolSizeH
+      );
+      return;
+    }
+    int batch;
+#pragma omp parallel for private(batch)
+    for(batch = 0; batch < numBatch; ++batch) {
+      fractional_max_pool3d_out_single_batch_frame<scalar_t>(
+        input + batch * numPlanes * inputH * inputW * inputT,
+        output + batch * numPlanes * outputH * outputW * outputT,
+        indices + batch * numPlanes * outputH * outputW * outputT,
+        randomSamples + batch * numPlanes * 3,
+        numPlanes,
+        inputT, inputW, inputH,
+        outputT, outputW, outputH,
+        poolSizeT, poolSizeW, poolSizeH
+      );
+    }
+  }
+
+void fractional_max_pool3d_out_cpu_template(
+  Tensor& output,
+  Tensor& indices,
+  const Tensor& input_,
+  IntList pool_size,
+  IntList output_size,
+  const Tensor& randomSamples) {
+
+  int outputT = output_size[0];
+  int outputW = output_size[1];
+  int outputH = output_size[2];
+  int poolSizeT = pool_size[0];
+  int poolSizeW = pool_size[1];
+  int poolSizeH = pool_size[2];
```

Comment:
Ordering here will need fixing (a test that takes tuple pool size will help you observe the problem)

Reply:
Fixed